modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheProcessor.java

```diff
@@ -5507,6 +5507,8 @@ private class CacheRecoveryLifecycle implements MetastorageLifecycleListener, Da
             CacheJoinNodeDiscoveryData data = locCfgMgr.restoreCacheConfigurations();
 
             cachesInfo.onStart(data);
+
+            ctx.query().registerMetadataForRegisteredCaches(false);
```

Comment:
Judging by your description, fix appears to be correct. Unfortunately, I'm also not the best person to say that, you know this code better than I do now :)